File: platforms/android/App/aj/src/main/java/applica/aj/runtime/websocket/AJWebSocketRuntime.java
package applica.aj.runtime.websocket;

import android.content.Context;
import android.util.Base64;
import android.util.DisplayMetrics;
import android.util.Log;

import org.apache.commons.lang3.StringUtils;
import org.json.JSONObject;

import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import applica.aj.AJObject;
import applica.aj.ManualSemaphore;
import applica.aj.Semaphore;
import applica.aj.runtime.AJRuntime;
import applica.aj.runtime.Buffer;
import applica.aj.runtime.Plugin;
import applica.framework.android.utils.CollectionUtils;
import io.socket.client.Ack;
import io.socket.client.IO;
import io.socket.client.Socket;
import io.socket.emitter.Emitter;

/**
 * Created by bimbobruno on 26/04/16.
 */
public class AJWebSocketRuntime extends AJRuntime {

    private boolean connecting;
    private Socket io;

    private List<ManualSemaphore> semaphores = new ArrayList<>();

    private static final ExecutorService EXECUTOR_SERVICE = Executors.newFixedThreadPool(10);

    public AJWebSocketRuntime(Context context, final URI uri) {
        super(context);

        IO.Options options = new IO.Options();
        options.reconnection = false;

        io = IO.socket(uri, options);

        io.on("trigger", new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                String type = (String) args[0];
                JSONObject json = (JSONObject) args[1];
                AJObject state = json != null ? AJObject.fromJson(json.toString()) : new AJObject();
                Ack ack = (Ack) args[2];

                try {
                    trigger(type, state);
                } catch (Throwable t) {
                    t.printStackTrace();
                }

                ack.call();
            }
        });

        io.on("exec", new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                String plugin = (String) args[0];
                String fn = (String) args[1];
                JSONObject json = (JSONObject) args[2];
                AJObject data = json != null ? AJObject.fromJson(json.toString()) : new AJObject();
                final Ack ack = (Ack) args[3];

                try {
                    exec(plugin, fn, data, new Plugin.Callback() {
                        @Override
                        public void onSuccess(AJObject data) {
                            ack.call(false, data);
                        }

                        @Override
                        public void onError(AJObject data) {
                            ack.call(true, data);
                        }
                    });
                } catch (Throwable t) {
                    t.printStackTrace();
                }

                ack.call();
            }
        });

        io.on("freeSemaphore", new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                int id = (int) args[0];
                freeSemaphore(id);
            }
        });

        io.on("device", new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                DisplayMetrics displayMetrics = getContext().getResources().getDisplayMetrics();

                AJObject device = AJObject.create()
                        .set("scale", displayMetrics.density)
                        .set("height", displayMetrics.heightPixels)
                        .set("width", displayMetrics.widthPixels);

                Ack ack = (Ack) args[0];
                ack.call(device.toJson());
            }
        });

        io.on("createBuffer", new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                Ack ack = (Ack) args[1];

                String base64 = (String) args[0];
                if (StringUtils.isEmpty(base64)) {
                    ack.call(true, "Bad data");
                    return;
                }

                byte[] bytes = Base64.decode(base64, Base64.DEFAULT);
                if (bytes == null) {
                    ack.call(true, "Bad data");
                    return;
                }

                ack.call(false, Buffer.create(bytes));
            }
        });

        io.on("readBuffer", new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                Ack ack = (Ack) args[1];

                Integer id = (Integer) args[0];
                if (id == null) {
                    ack.call(true, "Bad id");
                    return;
                }

                byte[] bytes = Buffer.get(id);
                if (bytes == null) {
                    ack.call(true, "Buffer not found: " + id.toString());
                    return;
                }

                String base64 = Base64.encodeToString(bytes, Base64.DEFAULT);

                ack.call(false, base64);
            }
        });

        io.on("deleteBuffer", new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                Ack ack = (Ack) args[1];

                Integer id = (Integer) args[0];
                if (id == null) {
                    ack.call(true, "Bad id");
                    return;
                }

                Buffer.destroy(id);

                ack.call(false);
            }
        });

        io.on(Socket.EVENT_CONNECT, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                connecting = false;
                Log.i("AJ", "Connected to websocket server");
            }
        });

        io.on(Socket.EVENT_CONNECTING, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                Log.i("AJ", "Connecting to websocket server on " + uri.toString());
                connecting = true;
            }
        });

        io.on(Socket.EVENT_CONNECT_ERROR, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                connecting = false;
                Log.w("AJ", "Error connecting to websocket server");
            }
        });

        io.on(Socket.EVENT_CONNECT_TIMEOUT, new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                connecting = false;
                Log.w("AJ", "Timeout connecting to websocket server");
            }
        });

        connecting = true;
        io.connect();
    }

    private void freeSemaphore(final int id) {
        ManualSemaphore semaphore = CollectionUtils.first(semaphores, new CollectionUtils.Predicate<ManualSemaphore>() {
            @Override
            public boolean evaluate(ManualSemaphore obj) {
                return obj.getId() == id;
            }
        });

        if (semaphore != null) {
            semaphore.green();
            semaphores.remove(semaphore);
        }
    }

    @Override
    public Semaphore run(String action, AJObject data) {
        data = data != null ? data : new AJObject();

        String json = data.toJson();

        while (connecting) {
            try {
                Log.i("AJ", "Waiting for connection with web socket server...");
                Thread.sleep(250);
            } catch (InterruptedException e) {}
        }

        Log.i("AJ", String.format("Calling %s with data: %s", action, json));

        final ManualSemaphore semaphore = new ManualSemaphore();
        semaphores.add(semaphore);

        io.emit("run", action, json, new Ack() {
            @Override
            public void call(Object... args) {
                freeSemaphore(semaphore.getId());
            }
        });

        return semaphore;

    }
}

File: platforms/ios/Libs/AJ/AJ/AJWebSocketRuntime.swift
 //
 //  AJWebSocketRuntime.swift
 //  AJ
 //
 //  Created by Bruno Fortunato on 17/02/16.
 //  Copyright © 2016 Bruno Fortunato. All rights reserved.
 //
 
 import UIKit
 import SocketIO
 
 open class AJWebSocketRuntime: AJRuntime {
    
    let io: SocketIOClient
    
    private var _connecting = false
    
    var semaphores = [AJManualSemaphore]()
    
    public init(url: URL) {
        let handleQueue = DispatchQueue(label: "AJ.AJWebSocketRuntime.handleQueue", attributes: DispatchQueue.Attributes.concurrent)
        io = SocketIOClient(socketURL: url, config: [.log(false), .handleQueue(handleQueue), .forceWebsockets(true), .reconnects(false)])
        super.init()
        
        io.on("trigger") { (data, ack) -> Void in
            let store = data[0] as! String
            let dict = data[1] as? [String: AnyObject]
            let argument: AJObject = dict != nil ? AJObject(dict: dict!) : AJObject.empty()
            
            self.tigger(store: store, data: argument)
            
            ack.with()
        }
        
        io.on("exec") { (data, ack) -> Void in
            let plugin = data[0] as! String
            let fn = data[1] as! String
            let dict = data[2] as? [String: AnyObject]
            let argument: AJObject = dict != nil ? AJObject(dict: dict!) : AJObject.empty()
            
            _ = self.exec(plugin: plugin, fn: fn, data: argument) { (error, result) in
                ack.with(error, result?.toJson() ?? "{}")
            }
        }
        
        io.on("device") { (data, ack) -> Void in
            let device = AJObject.create()
                .set("scale", Float(UIScreen.main.scale) as AnyObject?)
                .set("height", Int(UIScreen.main.bounds.height) as AnyObject?)
                .set("width", Int(UIScreen.main.bounds.width) as AnyObject?)
            
            
            ack.with(device.toJson())
        }
        
        io.on("freeSemaphore") { data, ack in
            let id = data[0] as! Int
            self.freeSemaphore(id)
        }
        
        io.on("createBuffer") { data, ack in
            guard let base64 = data[0] as? String else {
                ack.with(true, "Bad data")
                return
            }
            
            guard let bytes = Data(base64Encoded: base64, options: NSData.Base64DecodingOptions(rawValue: 0)) else {
                ack.with(true, "Bad data")
                return
            }
            
            let id = AJBuffer.create(with: bytes)
            ack.with(false, id)
        }
        
        io.on("readBuffer") { data, ack in
            guard let id = data[0] as? Int else {
                ack.with(true, "Bad data")
                return
            }
            
            guard let base64 = AJBuffer.get(id)?.base64EncodedData(options: Data.Base64EncodingOptions(rawValue: 0)) else {
                ack.with(true, "Bad data")
                return
            }
            
            ack.with(false, base64)
        }
        
        io.on("connect") { data, ack in
            print("Connected to web socket server")
            self._connecting = false
        }
        
        _connecting = true
        io.connect()
    }
    
    open override func run(action: String, data: AJObject = AJObject.empty()) -> AJSemaphore {
        let json = data.toJson();
        
        while _connecting {
            NSLog("Waiting for connection with web socket server...")
            Thread.sleep(forTimeInterval: 0.25)
        }
        
        NSLog("Calling \(action) with data: \(json)")
        
        let semaphore = AJManualSemaphore()
        semaphores.append(semaphore)

        self.io.emitWithAck("run", action, json).timingOut(after: 0) { (data) in
            self.freeSemaphore(semaphore.id)
        }
        
        return semaphore
    }
    
    func freeSemaphore(_ id: Int) {
        var index = -1
        var found = false
        for semaphore in semaphores {
            index += 1
            
            if semaphore.id == id {
                found = true
                semaphore.green()
                break;
            }
        }
        
        if found {
            semaphores.remove(at: index)
        }
    }
    
    open override func destroy() {
        super.destroy()
        
        semaphores = []
        io.disconnect()
    }
    
    
 }
